knora/dsplib/utils/xml_upload.py

```diff
@@ -843,65 +769,147 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
+                                                          resclass_name_2_type, id2iri_mapping, con, failed_uploads)
+    except BaseException as err:
+        handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+        exit(1)
+
+    # update the resources with the stashed XML texts
+    nonapplied_xml_texts = {}
+    if len(stashed_xml_texts) > 0:
+        try:
+            nonapplied_xml_texts = update_stashed_xml_texts(verbose, id2iri_mapping, con, stashed_xml_texts)
+        except BaseException as err:
+            handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+            exit(1)
+
+    # update the resources with the stashed resptrs
+    nonapplied_resptr_props = {}
+    if len(stashed_resptr_props) > 0:
+        try:
+            nonapplied_resptr_props = update_stashed_resptr_props(verbose, id2iri_mapping, con, stashed_resptr_props)
+        except BaseException as err:
+            handle_upload_error(err, input_file, id2iri_mapping, failed_uploads, stashed_xml_texts, stashed_resptr_props)
+            exit(1)
+
+    # write log files
+    exit_code = 0
+    timestamp_str = datetime.now().strftime("%Y%m%d-%H%M%S")
+    write_id2iri_mapping(input_file, id2iri_mapping, timestamp_str)
+    if len(nonapplied_xml_texts) > 0:
+        write_stashed_xml_texts(nonapplied_xml_texts, timestamp_str)
+        exit_code = 1
+    if len(nonapplied_resptr_props) > 0:
+        write_stashed_respr_props(nonapplied_resptr_props, timestamp_str)
+        exit_code = 1
 
-    # create a dictionary to look up permissions
-    permissions_lookup: dict[str, Permissions] = {}
-    for key, perm in permissions.items():
-        permissions_lookup[key] = perm.get_permission_instance()
+    if failed_uploads:
+        print(f"Could not upload the following resources: {failed_uploads}")
+        exit_code = 1
 
-    # create a dictionary to look up resource classes
-    resclass_name_2_type: dict[str, type] = {}
-    for res_class_name in project.get_resclass_names():
-        resclass_name_2_type[res_class_name] = project.get_resclass_type(res_class_name)
+    exit(exit_code)
 
-    res_iri_lookup: dict[str, str] = {}
 
-    failed_uploads = []
+def try_sipi_upload(sipi_server: Sipi, filepath: str) -> dict[Any, Any]:
+    """
+    Wrapper around sipi_server.upload_bitstream(filepath), that handles errors and retries in case of connectivity
+    issues.
+
+    Args:
+        sipi_server: Sipi instance
+        filepath: file to upload
+
+    Returns:
+        answer from knora
+    """
+
+    img = None
+    for _ in range(5):
+        try:
+            img = sipi_server.upload_bitstream(filepath)
+            break
+        except ConnectionError:
+            print(f'{datetime.now().isoformat()}: Try reconnecting to DSP server (SIPI)...')
+            time.sleep(1)
```

Comment:
might it maybe make more sense to sleep a bit longer? Now you sleep 5*1 seconds, and not necessarily all network issues recover within 5 seconds, right?

You could even have a growing timeout that uses the index as a loop:
```python
for t in range(5):
    ...
    sleep(2**i)
```
this would sleep for 1, 2, 4, 8 and 16 seconds (total 31 seconds)

Reply:
Don't you think there is a danger that big uploads that take weeks under normal circumstances will become even longer? Initially I wanted to sleep longer, but then this concern was raised. I find it difficult to say if the concern is a real danger.

Reply:
hard to say... but then, this should not delay the upload longer than the network outage lasts, right?